[Docs] Add docs for Sky Serve (#2794)
skypilot-org/skypilot: docs/source/examples/sky-serve.rst

    To see logs of load balancer:   sky serve logs --load-balancer sky-service-e4fb
    To see logs of controller:      sky serve logs --controller sky-service-e4fb

    To monitor replica status:      watch -n10 sky serve status sky-service-e4fb
    To send a test request:         curl -L <endpoint-url>

    SkyServe is spinning up your service now.
    The replicas should be ready within a short time.

Once any of the replicas becomes ready to serve, you can start sending requests to :code:`<endpoint-url>`. You can use :code:`watch -n10 sky serve status sky-service-e4fb` to monitor the latest status of the service. Once its status becomes :code:`READY`, you can start sending requests to :code:`<endpoint-url>`:

.. code-block:: console

    $ curl -L <endpoint-url>
    <html>
    <head>
        <title>My First Sky Serve Service</title>
    </head>
    <body>
        <p>Hello, Sky Serve!</p>
    </body>
    </html>

.. note::

  The :code:`curl` command won't follow the redirect and print the content of the redirected page by default. Since we are using HTTP-redirect, you need to use :code:`curl -L <endpoint-url>`.

> 

Author: If we are adding this, do you think we still need the vicuna example? Not sure if it is a little bit redundant if we include TGI...

> 

Reviewer: Some redundancy is fine. The main motivation is to make the very first impression about real, useful AI serving. Currently it's HTTP server.

How about we add a "Quickstart: TGI service" section (or change it to vLLM/FastChat etc.), but keep the user doc's concise formatting -- 1 snippet showing YAML, 1 snippet showing service status, then add 1 snippet showing how to CURL it correctly. With minimal text throughout.

.. _sky-serve:

> Sky Serve

Reviewer: Let's change all "Sky Serve" to "SkyServe" in both files?

=========

> Sky Serve is SkyPilot's serving library. Sky Serve takes an existing serving
> framework and deploys it across one or more regions or clouds.

Reviewer: ```suggestion
SkyServe (short for SkyPilot Serving) takes an existing serving
framework and deploys it across one or more regions or clouds.
```

You can find more configurations in :ref:`here <service-yaml-spec>`. This example will spin up two replicas of the service, each listening on port 8080. The service is considered ready when it responds to :code:`GET /health` with a 200 status code. You can customize the readiness probe by specifying a different path in the :code:`readiness_probe` field. By calling:

.. code-block:: console

    $ sky serve up hello-sky-serve.yaml

Sky Serve will start a centralized controller/load balancer and deploy the service to the cloud with the best price/performance ratio. It will also monitor the service status and re-launch a new replica if one of them fails.

Under the hood, :code:`sky serve up`:

#. Launches a controller which handles autoscaling, monitoring and load balancing;
#. Returns a Service Endpoint which will be used to accept traffic;
#. Meanwhile, the controller provisions replica VMs which later run the services;
#. Once any replica is ready, the requests sent to the Service Endpoint will be **HTTP-redirect** to one of the endpoint replicas.

After the controller is provisioned, you'll see the following in :code:`sky serve status` output:

> .. image:: ../images/sky-serve-status-output-provisioning.png

Reviewer: One problem to fix throughout this file: the para immediately after an image becomes too hard to read due to no vertical space.

Sky Serve Architecture
----------------------

.. image:: ../images/sky-serve-architecture.png
    :width: 800
    :align: center
    :alt: Sky Serve Architecture

Sky Serve has a centralized controller VM that manages the deployment of your service. Each service will have a process group to manage its replicas and route traffic to them.

It is composed of the following components:

#. **Controller**: The controller will monitor the status of the replicas and re-launch a new replica if one of them fails. It also autoscales the number of replicas if autoscaling config is set (see :ref:`Service YAML spec <service-yaml-spec>` for more information).
#. **Load Balancer**: The load balancer will route the traffic to all ready replicas. It is a lightweight HTTP server that listens on the service endpoint and **HTTP-redirects** the requests to one of the replicas.

All of the process group shares a single controller VM. The controller VM will be launched in the cloud with the best price/performance ratio. You can also :ref:`customize the controller resources <customizing-sky-serve-controller-resources>` based on your needs.

> An end-to-end LLM example

Reviewer: ```suggestion
Tutorial: Serve a Chatbot LLM!
```

-------------------------

> Below we show an end-to-end example of deploying a LLM model with Sky Serve. We'll use the `Vicuna OpenAI API Endpoint YAML <https://github.com/skypilot-org/skypilot/blob/master/llm/vicuna/serve-openai-api-endpoint.yaml>`_ as an example:

Reviewer: ```suggestion
Let’s bring up a real LLM chat service with FastChat + Vicuna. We'll use the `Vicuna OpenAI API Endpoint YAML <https://github.com/skypilot-org/skypilot/blob/master/llm/vicuna/serve-openai-api-endpoint.yaml>`_ as an example:
```

      echo 'Starting openai api server...'
      python -u -m fastchat.serve.openai_api_server --host 0.0.0.0 --port 8080 | tee ~/openai_api_server.log

    envs:
      MODEL_SIZE: 7

The above SkyPilot Task YAML will launch an OpenAI API endpoint with a 7B Vicuna model. This YAML can be used with :code:`sky launch` to launch a single replica of the service. By adding a :code:`service` section to the YAML, we can scale it into multiple replicas across multiple regions/clouds:

.. code-block:: yaml

    # vicuna.yaml
    service:
      readiness_probe: /v1/models
      replicas: 3

    resources:
      ports: 8080
      # Here goes other resources config

    # Here goes other task config

Now you have a Service YAML that can be used with Sky Serve! Simply run :code:`sky serve up vicuna.yaml -n vicuna` to deploy the service (use :code:`-n` to give your service a name!). After a while, there will be an OpenAI Compatible API endpoint ready to accept traffic (:code:`44.201.113.28:30001` in the following example):

.. image:: ../images/sky-serve-status-vicuna-ready.png
    :width: 800
    :align: center
    :alt: sky-serve-status-vicuna-ready

> 

Reviewer: We should add a CURL based command like the user doc. Does something like this work
```
Then it’s ready to accept traffic!

$ curl -L Y.Y.Y.Y:8082/generate \
    -X POST \
    -d '{"inputs":"What is Deep Learning?","parameters":{"max_new_tokens":20}}' \
    -H 'Content-Type: application/json'
{"generated_text":"\n nobody knows"}
```

Sky Serve
=========

Sky Serve is SkyPilot's serving library. Sky Serve takes an existing serving
framework and deploys it across one or more regions or clouds.

.. * Serve on scarce resources (e.g., A100; spot) with **reduced costs and increased availability**

Why Sky Serve?

* **Bring any serving framework** (vLLM, TGI, FastAPI, ...) and scale it across regions/clouds

> * **Reduce costs and increase availability** of service replicas by leveraging multiple/cheaper locations and hardware

Reviewer: ```suggestion
* **Reduce costs and increase availability** of service replicas by leveraging multiple/cheaper locations and hardware (spot instances)
```
